central/scannerdefinitions/handler/offline.go

```diff
@@ -0,0 +1,211 @@
+package handler
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	blob "github.com/stackrox/rox/central/blob/datastore"
+	"github.com/stackrox/rox/central/scannerdefinitions/file"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/protocompat"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+// offlineFileManager coordinates reads and writes to offline files that are backed
+// by the blob store.
+type offlineFileManager struct {
+	blobStore blob.Datastore
+
+	// dataDir the base directory where offline files will be written in prep for reads.
+	dataDir string
+
+	// files maps a blob's name to a file.
+	files map[string]*offlineFile
+}
+
+// offlineFileOpenerFunc returns a handle to a freshly opened file and its size.
+type offlineFileOpenerFunc func() (io.ReadCloser, int64, error)
+
+// newOfflineFileManager creates a new offline file manager.
+func newOfflineFileManager(blobStore blob.Datastore, dataDir string) *offlineFileManager {
+	return &offlineFileManager{
+		files:     map[string]*offlineFile{},
+		blobStore: blobStore,
+		dataDir:   dataDir,
+	}
+}
+
+// Register registers a blob name to be managed. A blob name must be registered before
+// the associated offline file can be used.
+func (o *offlineFileManager) Register(blobName string) {
+	o.files[blobName] = newOfflineFile(o.dataDir, o.blobStore, blobName)
+}
+
+// Open will return a handle for reading the offline file.
+// If the offline file was not found will return (nil, nil).
+func (o *offlineFileManager) Open(ctx context.Context, blobName string) (*vulDefFile, error) {
```

Comment:
wondering if instead of having a single manager which requires registration and putting in a blob name each time, we can have a factory which will give us a manager for each blob? Each manager is in charge of the lifecycle of its own blobstore/file. I think this would eliminate the need for registration and the map checks

Reply:
Attempted to create separate distinct objects in the handler, this works - however on upload/post will still need to dynamically find the correct offline file, which to implement would need either a series of conditions or a map.  

Keeping the PR as is, with the offline files inside the manager and doing the lookup by blob name has the advantage of being easier to test and requires minimal changes to add more offline files in the future.

Given that my vote is to keep this as is. Let me know if you feel strongly about it and can revise.